Add numeric_only=True to relevant pandas operations

blychs opened the pull request:
This pull request adds the keyword `numeric_only=True` to relevant pandas objects,
i.e. pandas.core.groupby.DataFrameGroupBy and pandas.core.window.rolling.Rolling .
The relevant operations are `mean()`, `min()`, `max()`, `median()` and `quantile()`.
This makes it compatible with pandas 2.2.2, and should be a nice start for future proofing MM, without breaking backwards compatibility (i., e., pandas 1, for monet).


If we want to keep it future proof, we should also update the references to freq='H' to freq='h', but I am not sure about compatibility with older versions of pandas for this, so I have kept it the way it was for now.

This has been tested for the surface plots with wrf-chem and AirNow.

Let me know if additional tests are required.

zmoon reviewed (commented):
@blychs I think you could go ahead with the change to using `'h'` for hourly freq instead of `'H'`. I tested using `freq='h'` in date range with pandas v1.0.5 (June 2020) and it worked fine.

Some places I identified:

* `melodies_monet/_cli.py` date ranges
* date range in `docs/examples/idealized.ipynb`
* `ts_avg_window` in `docs/examples/` and `examples/yaml` YAML files
* resamples in driver and `melodies_monet/util/tools.py`
* comments "pandas resample rule (e.g., 'H', 'D')" and "Pandas resampling rule (e.g., 'H', 'D')", which appear multiple places

The `reformat*` notebooks we are probably going to remove in the future so I wouldn't bother modifying them.

blychs commented:
@zmoon, I just prepared a new version, but now the newer scorecard plots (which I did not see there before) are failing in the updated version. I will update this pull request once I've corrected the issue. This pull request should not be accepted until I have fixed that.
Best
Pablo

blychs commented:
The corrections are done.  airnow_wrfchem.ipynb has the same result as the current code. The mistake was probably due to newer package versions, and is explained in the commit message. Let me know if any other issue appears.

Having said that, all the notebooks in docs/examples run without issues, except for camchem.ipynb. When I am running that, there is a conflict in the renaming of lat and lon. If I am not mistaken, though, the problem seems to be in the monet_accessor.py, since the dataset seems to have latitude and longitude created before renaming.

I might be wrong about that, though.

blychs commented:
So, @rschwant or @zmoon , the issue seems to have nothing to do with this PR. I added it as Issue # 261,](https://github.com/NOAA-CSL/MELODIES-MONET/issues/261#issue-2387428054) , together with a complete description and a few ideas of how to solve it. These can be at different levels, but I guess that that discussion should be continued there.

I have not found any other problems with this PR, and I think it is ready to undergo the tests that you deem necessary before (hopefully, if nothing else goes wrong) merging.

rschwant commented on melodies_monet/plots/surfplots.py:
```diff
@@ -1250,7 +1250,7 @@ def scorecard_step4_GetRegionLUCDate(ds_name=None,region_list=None,datelist=None
             region_date_urban_here = []
             for i in range(len(region_here['Time'])):
-                date_here1 = region_here['Time'][i]
+                date_here1 = region_here['Time'].values[i]
```
@btang1 can you confirm "values" can be added here? This seems like it should work fine, but was not in your original code.

zmoon replied:
Seems fine to me too.

btang1 replied:
@rschwant @zmoon I saw the change, I cannot say it works. we need to run and test, if test if past then we can change?
btw, who changed it?

rschwant reviewed (commented):
I just had 2 quick questions and then I can approve this later today. We can discuss quickly in the meeting today.

zmoon commented on melodies_monet/plots/surfplots.py:
```diff
@@ -43,7 +43,7 @@ def make_24hr_regulatory(df, col=None):
     df.index = df.time_local
     # select sites with nobs >=18, 75% completeness
-    df_24hr_ave = (df.groupby("siteid")[col].resample("D").sum(min_count=18)/df.groupby("siteid")[col].resample("D").count()).reset_index().dropna()
+    df_24hr_ave = (df.groupby("siteid")[col].resample("D").sum(min_count=18, numeric_only=True)/df.groupby("siteid")[col].resample("D").count()).reset_index().dropna()
```
count works for non-numeric data as well, so no issue.

rschwant replied:
Great then we can skip updating this one! Thanks for confirming. Can you confirm the values item below too? And then we are ready to merge this PR in.

rschwant reviewed (approved):
Zach confirmed my questions. Looks great thanks for updating this so we can move to Pandas 2!

rschwant commented:
@zmoon, Do you know why these tests are failing though? Is it using ts_avg_window: '3h' in the idealized example? What error have you all been seeing on this?

zmoon commented:
@rschwant the issue is an incompat between pandas v1 and matplotlib 3.9, which was released on conda-forge a week ago. pandas v2 [deals](https://github.com/pandas-dev/pandas/blame/ed09b58a1aff4814f437082e1dd0847f9a54e16f/pandas/plotting/_matplotlib/core.py#L873) with [the matplotlib change](https://matplotlib.org/stable/api/prev_api_changes/api_changes_3.7.0.html#legend-legendhandles) but I guess it wasn't backported.

To fix this for now, we could add `matplotlib<3.9` to the environment yml file.

rschwant commented:
Yes, that sounds good we can recommend an earlier version of matplotlib now too.

blychs commented:
Hi @rschwant , we saw the same with @dwfncar a couple of days ago.
See https://github.com/NOAA-CSL/MELODIES-MONET/issues/262 for some comments on that
Pablo

zmoon commented:
Thanks @blychs , reading your issue more closely would've saved me a few minutes. I made the change here in 352a088.

rschwant reviewed (approved):
Great thanks for making that update Zach!

rschwant commented:
Zach suggested it should be okay. Pablo added it in along with the other
updates to convert our code to be acceptable to Pandas 2.

It looks like our ReadTheDocs on the develop branch is still running okay
here:
https://melodies-monet.readthedocs.io/en/develop/examples/airnow_wrfchem.html

So it seems like you can go either way with this. Beiming let me know if it
does cause problems in your code though?

On Mon, Jul 15, 2024 at 11:58 PM Beiming Tang ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In melodies_monet/plots/surfplots.py
> <https://github.com/NOAA-CSL/MELODIES-MONET/pull/259#discussion_r1678800072>
> :
>
> > @@ -1250,7 +1250,7 @@ def scorecard_step4_GetRegionLUCDate(ds_name=None,region_list=None,datelist=None
>              region_date_rural_here = []
>              region_date_urban_here = []
>              for i in range(len(region_here['Time'])):
> -                date_here1 = region_here['Time'][i]
> +                date_here1 = region_here['Time'].values[i]
>
> @rschwant <https://github.com/rschwant> @zmoon <https://github.com/zmoon>
> I saw the change, I cannot say it works. we need to run and test, if test
> if past then we can change?
> btw, who changed it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-CSL/MELODIES-MONET/pull/259#discussion_r1678800072>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJD4V5WIFLKDH4HEPPORYPTZMSZANAVCNFSM6AAAAABJW3NSMCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCNZZGI4DANRRHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Becky Schwantes
Research Chemist
NOAA Chemical Sciences Laboratory
